File: priv/parsers/javascript/parser.js
#!/usr/bin/env node
/**
 * JavaScript/TypeScript AST Parser for RSOLV RFC-031
 * Uses Babel to parse JS/TS code and returns AST in JSON format via stdin/stdout
 */

const parser = require('@babel/parser');
const traverse = require('@babel/traverse').default;
const t = require('@babel/types');

// Set up timeout handler
process.on('SIGALRM', () => {
    const errorResponse = {
        status: "error",
        error: {
            type: "TimeoutError",
            message: "Parser timeout after 30 seconds"
        }
    };
    console.log(JSON.stringify(errorResponse));
    process.exit(1);
});

function nodeToDict(node) {
    /**
     * Convert Babel AST node to dictionary format with circular reference handling
     */
    if (node === null || node === undefined) {
        return null;
    }
    
    if (typeof node !== 'object') {
        return node;
    }
    
    if (Array.isArray(node)) {
        return node.map(item => nodeToDict(item));
    }
    
    const result = {
        type: node.type
    };
    
    // Add location info if available
    if (node.loc) {
        result._loc = {
            start: {
                line: node.loc.start.line,
                column: node.loc.start.column
            },
            end: {
                line: node.loc.end.line,
                column: node.loc.end.column
            }
        };
    }
    
    if (node.start !== undefined) {
        result._start = node.start;
    }
    
    if (node.end !== undefined) {
        result._end = node.end;
    }
    
    // Process all properties except circular references
    for (const key in node) {
        if (key === 'parent' || key === 'scope' || key === 'hub' || key === '_parent') {
            // Skip circular references
            continue;
        }
        
        if (node.hasOwnProperty(key) && key !== 'type' && key !== 'loc' && key !== 'start' && key !== 'end') {
            result[key] = nodeToDict(node[key]);
        }
    }
    
    return result;
}

function findSecurityPatterns(ast) {
    /**
     * Extract security-relevant patterns from Babel AST
     */
    const patterns = [];
    
    traverse(ast, {
        // Dangerous function calls
        CallExpression(path) {
            const node = path.node;
            let funcName = '';
            
            if (t.isIdentifier(node.callee)) {
                funcName = node.callee.name;
            } else if (t.isMemberExpression(node.callee)) {
                if (t.isIdentifier(node.callee.object) && t.isIdentifier(node.callee.property)) {
                    funcName = `${node.callee.object.name}.${node.callee.property.name}`;
                } else if (t.isIdentifier(node.callee.property)) {
                    funcName = node.callee.property.name;
                }
            }
            
            // Common dangerous functions
            const dangerousFuncs = new Set([
                'eval', 'Function', 'setTimeout', 'setInterval', 
                'document.write', 'innerHTML', 'outerHTML',
                'exec', 'spawn', 'execSync'
            ]);
            
            if (dangerousFuncs.has(funcName) || funcName.includes('eval') || funcName.includes('innerHTML')) {
                patterns.push({
                    type: 'dangerous_function',
                    function: funcName,
                    line: node.loc ? node.loc.start.line : 0,
                    column: node.loc ? node.loc.start.column : 0
                });
            }
        },
        
        // Template literals that might build SQL/commands
        TemplateLiteral(path) {
            const node = path.node;
            patterns.push({
                type: 'template_literal',
                line: node.loc ? node.loc.start.line : 0,
                column: node.loc ? node.loc.start.column : 0
            });
        },
        
        // String concatenation that might build queries
        BinaryExpression(path) {
            const node = path.node;
            if (node.operator === '+') {
                patterns.push({
                    type: 'string_concat',
                    line: node.loc ? node.loc.start.line : 0,
                    column: node.loc ? node.loc.start.column : 0
                });
            }
        },
        
        // Assignment to innerHTML or similar dangerous properties
        AssignmentExpression(path) {
            const node = path.node;
            if (t.isMemberExpression(node.left) && t.isIdentifier(node.left.property)) {
                const propName = node.left.property.name;
                if (['innerHTML', 'outerHTML', 'document.write'].includes(propName)) {
                    patterns.push({
                        type: 'dangerous_assignment',
                        property: propName,
                        line: node.loc ? node.loc.start.line : 0,
                        column: node.loc ? node.loc.start.column : 0
                    });
                }
            }
        }
    });
    
    return patterns;
}

function main() {
    /**
     * Main parser loop - reads JSON requests from stdin, writes responses to stdout
     */
    
    // Ensure stdout is unbuffered
    process.stdout.setEncoding('utf8');
    
    const readline = require('readline');
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        terminal: false
    });
    
    rl.on('line', (line) => {
        let timeout;
        try {
            // Set 30-second timeout for each request
            timeout = setTimeout(() => {
                const errorResponse = {
                    status: "error",
                    error: {
                        type: "TimeoutError",
                        message: "Parser timeout after 30 seconds"
                    }
                };
                console.log(JSON.stringify(errorResponse));
                process.exit(1);
            }, 30000);
            
            const request = JSON.parse(line.trim());
            const requestId = request.id || 'unknown';
            const command = request.command || '';
            const action = request.action || command; // Support both formats
            
            let response;
            
            if (action === 'HEALTH_CHECK') {
                response = {
                    id: requestId,
                    result: 'ok'
                };
            } else if (action !== 'parse' && command !== '') {
                // Handle command-based interface (for compatibility with PortWorker)
                const startTime = Date.now();
                const code = command; // Treat command as code to parse
                const options = request.options || {};
                const filename = request.filename || '<string>';
                
                // Determine if this is TypeScript based on filename or content
                const isTypeScript = filename.endsWith('.ts') || filename.endsWith('.tsx') || 
                                   code.includes('interface ') || code.includes('type ') || 
                                   code.includes(': string') || code.includes(': number');
                
                const language = isTypeScript ? 'typescript' : 'javascript';
                
                // Set up Babel parser options
                const parserOptions = {
                    sourceType: 'module',
                    allowImportExportEverywhere: true,
                    allowReturnOutsideFunction: true,
                    plugins: [
                        'jsx',
                        'asyncGenerators',
                        'decorators-legacy',
                        'doExpressions',
                        'exportDefaultFrom',
                        'exportNamespaceFrom',
                        'functionBind',
                        'objectRestSpread',
                        'dynamicImport'
                    ]
                };
                
                if (isTypeScript) {
                    parserOptions.plugins.push('typescript');
                }
                
                // Parse the code
                const ast = parser.parse(code, parserOptions);
                
                // Convert AST to dictionary
                const astDict = nodeToDict(ast);
                
                // Extract security patterns if requested
                let securityPatterns = [];
                if (options.include_security_patterns !== false) {
                    securityPatterns = findSecurityPatterns(ast);
                }
                
                const parseTimeMs = Date.now() - startTime;
                
                // Count nodes by traversing AST
                let nodeCount = 0;
                traverse(ast, {
                    enter() {
                        nodeCount++;
                    }
                });
                
                response = {
                    id: requestId,
                    status: 'success',
                    success: true,
                    ast: astDict,
                    security_patterns: securityPatterns,
                    metadata: {
                        parser_version: '1.0.0',
                        language: language,
                        language_version: process.version,
                        parse_time_ms: parseTimeMs,
                        ast_node_count: nodeCount
                    }
                };
            } else if (action !== 'parse' && !command) {
                response = {
                    id: requestId,
                    status: 'error',
                    error: {
                        type: 'InvalidAction',
                        message: `Unknown action: ${action}`
                    }
                };
            } else {
                const startTime = Date.now();
                const code = request.code || '';
                const options = request.options || {};
                const filename = request.filename || '<string>';
                
                // Determine if this is TypeScript
                const isTypeScript = filename.endsWith('.ts') || filename.endsWith('.tsx') || 
                                   code.includes('interface ') || code.includes('type ') || 
                                   code.includes(': string') || code.includes(': number');
                
                const language = isTypeScript ? 'typescript' : 'javascript';
                
                // Set up Babel parser options
                const parserOptions = {
                    sourceType: 'module',
                    allowImportExportEverywhere: true,
                    allowReturnOutsideFunction: true,
                    plugins: [
                        'jsx',
                        'asyncGenerators',
                        'decorators-legacy',
                        'doExpressions',
                        'exportDefaultFrom',
                        'exportNamespaceFrom',
                        'functionBind',
                        'objectRestSpread',
                        'dynamicImport'
                    ]
                };
                
                if (isTypeScript) {
                    parserOptions.plugins.push('typescript');
                }
                
                // Parse the code
                const ast = parser.parse(code, parserOptions);
                
                // Convert AST to dictionary
                const astDict = nodeToDict(ast);
                
                // Extract security patterns if requested
                let securityPatterns = [];
                if (options.include_security_patterns !== false) {
                    securityPatterns = findSecurityPatterns(ast);
                }
                
                const parseTimeMs = Date.now() - startTime;
                
                // Count nodes
                let nodeCount = 0;
                traverse(ast, {
                    enter() {
                        nodeCount++;
                    }
                });
                
                response = {
                    id: requestId,
                    status: 'success',
                    ast: astDict,
                    security_patterns: securityPatterns,
                    metadata: {
                        parser_version: '1.0.0',
                        language: language,
                        language_version: process.version,
                        parse_time_ms: parseTimeMs,
                        ast_node_count: nodeCount
                    }
                };
            }
            
            clearTimeout(timeout);
            console.log(JSON.stringify(response));
            
        } catch (error) {
            let requestId = 'unknown';
            try {
                const request = JSON.parse(line.trim());
                requestId = request.id || 'unknown';
            } catch (e) {
                // Could not parse request, use unknown ID
            }
            
            if (timeout) clearTimeout(timeout);
            
            let response;
            
            if (error.name === 'SyntaxError' || error.message.includes('Unexpected token')) {
                response = {
                    id: requestId,
                    status: 'error',
                    success: false,
                    error: {
                        type: 'SyntaxError',
                        message: error.message,
                        line: error.loc ? error.loc.line : undefined,
                        column: error.loc ? error.loc.column : undefined
                    }
                };
            } else {
                response = {
                    id: requestId,
                    status: 'error',
                    success: false,
                    error: {
                        type: error.constructor.name,
                        message: error.message
                    }
                };
            }
            
            console.log(JSON.stringify(response));
        }
    });
    
    rl.on('close', () => {
        process.exit(0);
    });
}

if (require.main === module) {
    main();
}

File: priv/parsers/python/parser.py
#!/usr/bin/env python3
"""
Python AST Parser for RSOLV RFC-031
Parses Python code and returns AST in JSON format via stdin/stdout
"""

import sys
import json
import ast
import time
import os
import signal

# Set up signal handler for timeout
def timeout_handler(signum, frame):
    error_response = {
        "status": "error",
        "error": {
            "type": "TimeoutError",
            "message": "Parser timeout after 30 seconds"
        }
    }
    print(json.dumps(error_response))
    sys.stdout.flush()
    sys.exit(1)

signal.signal(signal.SIGALRM, timeout_handler)

def node_to_dict(node):
    """Convert Python AST node to dictionary format"""
    if isinstance(node, ast.AST):
        fields = {}
        for field, value in ast.iter_fields(node):
            fields[field] = node_to_dict(value)
        
        result = {
            'type': node.__class__.__name__
        }
        
        # Add fields directly to result
        for field, value in fields.items():
            result[field] = value
        
        # Add location info if available
        if hasattr(node, 'lineno'):
            result['_lineno'] = node.lineno
        if hasattr(node, 'col_offset'):
            result['_col_offset'] = node.col_offset
        if hasattr(node, 'end_lineno'):
            result['_end_lineno'] = node.end_lineno
        if hasattr(node, 'end_col_offset'):
            result['_end_col_offset'] = node.end_col_offset
            
        return result
    elif isinstance(node, list):
        return [node_to_dict(item) for item in node]
    else:
        return node

def find_security_patterns(tree):
    """Extract security-relevant patterns from AST"""
    patterns = []
    
    class SecurityVisitor(ast.NodeVisitor):
        def visit_Call(self, node):
            # Look for dangerous function calls
            func_name = None
            if isinstance(node.func, ast.Name):
                func_name = node.func.id
            elif isinstance(node.func, ast.Attribute):
                func_name = node.func.attr
            
            # Common dangerous functions
            dangerous_funcs = {
                'eval', 'exec', 'compile', '__import__',
                'subprocess.call', 'subprocess.run', 'os.system',
                'open', 'file'
            }
            
            if func_name in dangerous_funcs:
                patterns.append({
                    'type': 'dangerous_function',
                    'function': func_name,
                    'line': node.lineno,
                    'column': node.col_offset
                })
            
            self.generic_visit(node)
        
        def visit_JoinedStr(self, node):
            # f-strings that might be building SQL/commands
            patterns.append({
                'type': 'f_string',
                'line': node.lineno,
                'column': node.col_offset
            })
            self.generic_visit(node)
        
        def visit_BinOp(self, node):
            # String concatenation that might build queries
            if isinstance(node.op, ast.Add):
                if (isinstance(node.left, ast.Str) or 
                    isinstance(node.right, ast.Str)):
                    patterns.append({
                        'type': 'string_concat',
                        'line': node.lineno,
                        'column': node.col_offset
                    })
            self.generic_visit(node)
    
    visitor = SecurityVisitor()
    visitor.visit(tree)
    
    return patterns

def main():
    """Main parser loop - reads JSON requests from stdin, writes responses to stdout"""
    
    # Ensure stdout is unbuffered
    sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 1)
    
    while True:
        try:
            line = sys.stdin.readline()
            if not line:
                break
            
            # Set 30-second timeout for each request
            signal.alarm(30)
            
            request = json.loads(line.strip())
            request_id = request.get('id', 'unknown')
            command = request.get('command', '')
            action = request.get('action', command)  # Support both formats
            
            if action == 'HEALTH_CHECK':
                response = {
                    'id': request_id,
                    'result': 'ok'
                }
            elif action != 'parse' and command != '':
                # Handle command-based interface (for compatibility with PortWorker)
                start_time = time.time()
                code = command  # Treat command as code to parse
                options = request.get('options', {})
                filename = request.get('filename', '<string>')
                
                # Parse the code
                tree = ast.parse(code, filename=filename)
                
                # Convert AST to dictionary
                ast_dict = node_to_dict(tree)
                
                # Extract security patterns if requested
                security_patterns = []
                if options.get('include_security_patterns', True):
                    security_patterns = find_security_patterns(tree)
                
                parse_time_ms = int((time.time() - start_time) * 1000)
                
                response = {
                    'id': request_id,
                    'status': 'success',
                    'success': True,
                    'ast': ast_dict,
                    'security_patterns': security_patterns,
                    'metadata': {
                        'parser_version': '1.0.0',
                        'language': 'python',
                        'language_version': sys.version.split()[0],
                        'parse_time_ms': parse_time_ms,
                        'ast_node_count': len(list(ast.walk(tree)))
                    }
                }
            elif action != 'parse' and not command:
                response = {
                    'id': request_id,
                    'status': 'error',
                    'error': {
                        'type': 'InvalidAction',
                        'message': f"Unknown action: {action}"
                    }
                }
            else:
                start_time = time.time()
                code = request.get('code', '')
                options = request.get('options', {})
                filename = request.get('filename', '<string>')
                
                # Parse the code
                tree = ast.parse(code, filename=filename)
                
                # Convert AST to dictionary
                ast_dict = node_to_dict(tree)
                
                # Extract security patterns if requested
                security_patterns = []
                if options.get('include_security_patterns', True):
                    security_patterns = find_security_patterns(tree)
                
                parse_time_ms = int((time.time() - start_time) * 1000)
                
                response = {
                    'id': request_id,
                    'status': 'success',
                    'ast': ast_dict,
                    'security_patterns': security_patterns,
                    'metadata': {
                        'parser_version': '1.0.0',
                        'language': 'python',
                        'language_version': sys.version.split()[0],
                        'parse_time_ms': parse_time_ms,
                        'ast_node_count': len(list(ast.walk(tree)))
                    }
                }
                
        except SyntaxError as e:
            response = {
                'id': request_id,
                'status': 'error',
                'success': False,
                'error': {
                    'type': 'SyntaxError',
                    'message': str(e),
                    'line': e.lineno,
                    'offset': e.offset,
                    'text': e.text
                }
            }
        except json.JSONDecodeError as e:
            response = {
                'id': 'unknown',
                'status': 'error',
                'success': False,
                'error': {
                    'type': 'JSONDecodeError',
                    'message': f"Invalid JSON: {str(e)}"
                }
            }
        except Exception as e:
            response = {
                'id': request_id,
                'status': 'error',
                'success': False,
                'error': {
                    'type': type(e).__name__,
                    'message': str(e)
                }
            }
        finally:
            # Cancel timeout
            signal.alarm(0)
        
        # Send response
        print(json.dumps(response))
        sys.stdout.flush()

if __name__ == '__main__':
    main()